apps/web/src/libs/event-emitter.ts

```diff
@@ -0,0 +1,35 @@
+export interface Subscriber {
+  (...args: any[]): void
+}
+
+export class EventEmitter {
+  private _subscribers: { [event: string]: Subscriber[] } = {}
+
+  public subscribe(event: string, function_: Subscriber) {
+    if (!this._subscribers[event]) {
+      this._subscribers[event] = []
+    }
+    this._subscribers[event].push(function_)
+    return () => {
+      this.unsubscribe(event, function_)
+    }
+  }
+
+  public unsubscribe(event: string, function_: Subscriber) {
+    if (this._subscribers[event]) {
+      this._subscribers[event] = this._subscribers[event].filter(
+        (subscriber) => subscriber !== function_
+      )
+    }
+  }
+
+  public emit(event: string, ...args: unknown[]) {
+    if (this._subscribers[event]) {
+      for (const subscriber of this._subscribers[event]) {
+        setTimeout(() => {
```

Comment:
Curious why we need a timeout. Is it just to bump it to the bottom of the callstack?

Reply:
In Node.js this would've been `setImmediate`, but that's not an official JS API. There is a polyfill but didn't wanna add that just for this. But yes, the idea is to push it to the bottom of the callstack.